Feat/coinbase advanced trade v2

MementoRC opened the pull request:
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
Coinbase Advanced Trade connector following v2.1 standard


**Tests performed by the developer**:
Passed all tests locally


**Tips for QA testing**:

nikspz commented:
hi @MementoRC Could you please add update related to the passphrase? 

Seems that new version of API keys not need passphrase to connect, but it's still required on this PR and also unable to connect with empty passphrase.


![image](https://github.com/hummingbot/hummingbot/assets/83953535/55617d34-7758-498f-b5b6-831f2e33e82b)

MementoRC commented:
@nikspz This connector cannot be used with Coinbase Pro/Exchange. Pro was discontinued but Exchange (institutional) used the same API, which may need to be refreshed.

Coinbase Advanced Trade is a NEW API and product from Coinbase, most non-institutional user of hummingbot used coinbase_pro accounts and are de facto going to need the Advanced Trade since thier accounts were converted into a Advanced Trade account, this is why I did not emphasize the distinction, but Hummingbot will likely need to support both

The Auth update solves the connection issue with Advanced Trade (Pro/Exchange has not changed, afaik), which was a timestamp passed as an `int` rather than `str`, their documentation was unclear to me

MementoRC commented:
How do I continue to update this codebase, I think there is a misconception on my part in terms of public/private calls, it seems that Hummingbot expects some calls to be public, but all the Advanced Trade calls are private, so I will likely need to change some implementation

nikspz commented:
commit [0feaf30](https://github.com/hummingbot/hummingbot/pull/6458/commits/0feaf3043f759f005912aa60c4d94a783def3715)
Tests performed:
- connect coinbase_advanced_trade_v2: ok
- checked balance: ok
- create strategy: ok
- start strategy: Failed, Client crashed and unresponsive
- restarted Client, import strategy: ok
- start strategy: Failed, Client crashed and not responding


![coinbasenew-client crashing](https://github.com/hummingbot/hummingbot/assets/83953535/380be037-40ef-4752-b3e7-7db29fde212e)

MementoRC commented:
@nikspz Thank you for the tirst run through. The startup error cannot be avoided, the trading rules infor from the exchange is a private call, thus is not authorized.
I probably should put more logging to better understand what is wrong
Thank you

MementoRC commented:
@nikspz, Thank you for looking into this, but the debug helper I added is not behaving as expected in all situation (just in mine where it finds the location where I want to write the debug file). This is annoying because I need the debug file (too many other error messages pop up in the normal log) to continue to fix the websocket connection, which needs more re-factoring.

The previous issue is fixed, signing works, orderbook works (only the buy/sell has not been fully tested). The issue with the websocket is that all the messages are bundled under one counter, so the sequence cannot be verified and we cannot know whether some messages have been lost. The re-factor is separating every channel/pair to its own websocket connection

I will continue to commit to save my progress, but it will, most likely, not be testable - I don't want to spend to much time on the DebugToFile since I don't think it will be accepted

I did not think you would spend time on this commit since it failed the check, please only look at the commits passing checks

Thank you

MementoRC commented:
@cardosofede , @nikspz I fixed the failing tests, a short QA on operations would be nice. I anticipate it not to work quite right:
   - Websocket should complain about out-of-sequence messages (The proposed solution is under testing)
   - DebugToFile creates a coinbase specific debug logfile under the `logs/` directory
 
It would be great to have a light QA to check that the Auth is now working correctly and find live issues with REST

cardosofede commented:
@MementoRC do you want QA to review the AUTH? I think that it's better to review all at once since if you make some changes in the auth module later on QA will need to retest.
Also, you can test that it's working by yourself and when it's ready QA can fully test the connector

nikspz commented:
Test performed:
- commit a138ba54a387225a5c23118fda17dbb9e782e3b6
    - Initial start: 
![image](https://github.com/hummingbot/hummingbot/assets/83953535/0b75c1fe-d9f3-4b53-b3fe-cf02e8be80c1)
```
2023-08-01 19:26:11,846 - 80968 - hummingbot.connector.exchange.coinbase_advanced_trade_v2.coinbase_advanced_trade_v2_exchange.CoinbaseAdvancedTradeV2Exchange - ERROR - Error getting all trading pairs from Coinbase Advanced Trade: Error executing request GET https://api.coinbase.com/api/v3/brokerage/products. HTTP status is 401. Error: Unauthorized
Traceback (most recent call last):
  File "/home/nikita/coinbasev2/hummingbot/connector/exchange/coinbase_advanced_trade_v2/coinbase_advanced_trade_v2_exchange.py", line 453, in _initialize_market_assets
    products: Dict[str, Any] = await self._api_get(
  File "/home/nikita/coinbasev2/hummingbot/connector/exchange_py_base.py", line 865, in _api_get
    return await self._api_request(*args, **kwargs)
  File "/home/nikita/coinbasev2/hummingbot/connector/exchange_py_base.py", line 907, in _api_request
    request_result = await rest_assistant.execute_request(
  File "/home/nikita/coinbasev2/hummingbot/core/web_assistant/rest_assistant.py", line 75, in execute_request
    raise IOError(f"Error executing request {method.name} {url}. HTTP status is {response.status}. "
OSError: Error executing request GET https://api.coinbase.com/api/v3/brokerage/products. HTTP status is 401. Error: Unauthorized
```

- connected coinbase advanced trade v2 successfully
- checked balance successfully 
- Created pureMM using coinbase_advanced_trade_v2 successfully
- Started strategy: Failed to place order

[logs_coin4.log](https://github.com/hummingbot/hummingbot/files/12233353/logs_coin4.log)
[commit.zip](https://github.com/hummingbot/hummingbot/files/12233352/commit.zip)
```
2023-08-01 19:30:00,993 - 80968 - hummingbot.core.rate_oracle.rate_oracle - INFO - Network status has changed to NetworkStatus.CONNECTED. Starting networking...
2023-08-01 19:30:01,000 - 80968 - hummingbot.core.clock - ERROR - Unexpected error running clock tick.
Traceback (most recent call last):
  File "hummingbot/core/clock.pyx", line 119, in hummingbot.core.clock.Clock.run_til
    child_iterator.c_tick(self._current_tick)
  File "hummingbot/strategy/pure_market_making/pure_market_making.pyx", line 750, in hummingbot.strategy.pure_market_making.pure_market_making.PureMarketMakingStrategy.c_tick
    self.c_apply_budget_constraint(proposal)
  File "hummingbot/strategy/pure_market_making/pure_market_making.pyx", line 941, in hummingbot.strategy.pure_market_making.pure_market_making.PureMarketMakingStrategy.c_apply_budget_constraint
    quote_size = buy.size * buy.price * (Decimal(1) + buy_fee.percent)
TypeError: unsupported operand type(s) for +: 'decimal.Decimal' and 'TradeFeeSchema'
2023-08-01 19:30:02,000 - 80968 - hummingbot.core.clock - ERROR - Unexpected error running clock tick.
Traceback (most recent call last):
  File "hummingbot/core/clock.pyx", line 119, in hummingbot.core.clock.Clock.run_til
    child_iterator.c_tick(self._current_tick)
  File "hummingbot/strategy/pure_market_making/pure_market_making.pyx", line 750, in hummingbot.strategy.pure_market_making.pure_market_making.PureMarketMakingStrategy.c_tick
    self.c_apply_budget_constraint(proposal)
  File "hummingbot/strategy/pure_market_making/pure_market_making.pyx", line 941, in hummingbot.strategy.pure_market_making.pure_market_making.PureMarketMakingStrategy.c_apply_budget_constraint
    quote_size = buy.size * buy.price * (Decimal(1) + buy_fee.percent)
TypeError: unsupported operand type(s) for +: 'decimal.Decimal' and 'TradeFeeSchema'
2023-08-01 19:30:03,000 - 80968 - hummingbot.core.clock - ERROR - Unexpected error running clock tick.
Traceback (most recent call last):
  File "hummingbot/core/clock.pyx", line 119, in hummingbot.core.clock.Clock.run_til
    child_iterator.c_tick(self._current_tick)
  File "hummingbot/strategy/pure_market_making/pure_market_making.pyx", line 750, in hummingbot.strategy.pure_market_making.pure_market_making.PureMarketMakingStrategy.c_tick
    self.c_apply_budget_constraint(proposal)
  File "hummingbot/strategy/pure_market_making/pure_market_making.pyx", line 941, in hummingbot.strategy.pure_market_making.pure_market_making.PureMarketMakingStrategy.c_apply_budget_constraint
    quote_size = buy.size * buy.price * (Decimal(1) + buy_fee.percent)
TypeError: unsupported operand type(s) for +: 'decimal.Decimal' and 'TradeFeeSchema'
```

![image](https://github.com/hummingbot/hummingbot/assets/83953535/f76bef37-e1b0-43e4-a58d-f0cbfc65a010)

MementoRC commented:
@nikspz Thank you for your time and help on this. The fee failure was unexpected. Back to the drawing board

nikspz commented:
hi @MementoRC Could you please give some updates on the bounty status? thanks

MementoRC commented:
@nikspz Apologies, I have been detracted away from hummingbot. I am struggling a bit with the OrderBookDataSource tests, I am debugging to find the piece that is not fitting quite correctly.

nikspz commented:
@MementoRC Could you please give some updates on the bounty status here? Thank you

MementoRC commented:
@nikspz , I made some progress, but a few tests in the orderbook stream are failing by not terminating (hang forever) and I have a hard time making progress on those

I should be able to focus more on hummingbot dev in the coming days

MementoRC commented:
@nikspz , local live test do not show errors, however, currently my account cannot place buys

ajjcoppola commented:
@MementoRC I'd like to test this.  I have the latest development version checked out.

MementoRC commented:
@ajjcoppola , great, I think you could use `gh pr checkout 6458` in your local dev directory. You may need to install `gh` with `conda install gh` and set the default remote to `hummingbot/hummingbot`

Note that the QA team has not yet re-tested after I fixed a few issues and my account has a `buy disabled` for unknown reasons, so my live tests were very limited: USE WITH CAUTION
Thank you

ajjcoppola commented:
@MementoRC Hve updated my repo was you suggested...have signed intomy coinbase account...now need to figure out what flavor of api this is...websockets...rest...fix...probably websockets, right?
You could point me to the tests you are running, or I'll look them up myself.  Thanks!

MementoRC commented:
@ajjcoppola So this is not an implementation of the full Coinbase Advanced trade API, but only the subset needed by hummingbot. It implements both the REST and the Websocket needed by Hummingbot. Coinbase Advanced Trade does not provide FIX

The tests are under `tests/hummingbot/connectors/exchange/coinbase_advanced_trade_v2`. Note that there are no live tests as this is typically performed by the QA team

There is a closed PR that had some testing based on web-crawling the documentation, but the framework was too different to be safely integrated: https://github.com/hummingbot/hummingbot/pull/6307. It does have a full REST API implementation

ajjcoppola commented:
There is a 
test/hummingbot/connector/exchange/coinbase_pro
is that what you mean?
Also, my API key won't be active till Sunday(they make you wait 48hrs it seems), so slight delay there...yet I'll put it all under my debugger and follow the flow  line by line that way!

ajjcoppola commented:
@MementoRC  So, I updated to python 3.10, reinstalled, and get by some errors...now get one saying I'm not authorized...it's just going through the connect code and trying to get the trading pairs.  Is the url right??? especially the /v3/ , which needs auth, which I believe I have supplied, and it is in the request.

builtins.OSError: Error executing request GET https://api.coinbase.com/api/v3/brokerage/products. HTTP status is 401. Error: Unauthorized

I'm in the debugger.  The next thing I can try is to see if I can do this request outside of hummingbot...just with curl...what do you think?

MementoRC commented:
@ajjcoppola You have to make sure that you are using Coinbase Advanced Trade and have a API key for that exchange. CoinbasePro is a different exchange. Everything for the Advanced is under `coinbase_advanced_trade_v2` subdirs.

It is possible that the `products` call throw an unauthorized because  hummingbot makes a general call to exchange information, but in the case of Coinbase Advanced Trade, the `products` is a authenticated call, thus the first call fails.

You should try to connect and type `balance` as your first hummingbot command, this should show you your balance on Coinbase Advanced Trade and verify that the API key is correct, if that fails, then there is likely an issue with your API key. You can use `Ctrl+Shift+V` to paste your key in the console (have you checked the discord channel?)

a curl request may be difficult, you need to create the signature to send to authenticate the API request, that can be tricky. I tested the my keys and they worked last time I tried.

zethereumGmail commented:
I tried to test the new coinbase advanced trade connector with a basic script and a pure_market_making strategy both failed.
I can connect and run balance to obtain a balance. 

When I start the basic pure_market_making strategy I just get the following output

Please let me know how I can assist in troubleshooting this. I am not a python programmer but have been a linux engineer for 20+ years and do code when necessary.


![hummingbot_coinbase_feature_branch_error](https://github.com/hummingbot/hummingbot/assets/69935809/043b15c1-2c9f-4036-9867-3add74d6a7d1)

zethereumGmail commented:
> @MementoRC So, I updated to python 3.10, reinstalled, and get by some errors...now get one saying I'm not authorized...it's just going through the connect code and trying to get the trading pairs. Is the url right??? especially the /v3/ , which needs auth, which I believe I have supplied, and it is in the request.
> 
> builtins.OSError: Error executing request GET https://api.coinbase.com/api/v3/brokerage/products. HTTP status is 401. Error: Unauthorized
> 
> I'm in the debugger. The next thing I can try is to see if I can do this request outside of hummingbot...just with curl...what do you think?

Are you locking down your API key by IP address? If so try finding out your IPv4 and IPv6 address and adding two entries. Coinbase doesn't seem to like putting two whitelisted IPs on the same token, at least it's never worked for me so I just create two. 
I have seen a similar error and it was fixed by adding my IPv4 address.

MementoRC commented:
@zethereumGmail Welcome onboard, wow, I better be on my best behavior moving forward. So, nothing points to an issue with the Coinbase Advanced Trade connector yet. The error you get is related to Bybit

The Message loss is due to a slight out-of-order connection that occurs with the heartbeats subscription because we are not actually listening to that channel, so we subscribe to it even though we have not started the listening task manager. Perhaps more information should be added to that warning, because we still want to be warned IF a real subscription is made before listening, which would incur loss of the first few messages until the listener starts (in the case of order book that would be the full order book snapshot sent as the first message)

zethereumGmail commented:
I am running in a Docker container and I have conf_client.yml log_level: DEBUG but I just have a 0 byte errors.log, I do have other files generated. The logs_hummingbot.log never mentions coinbase just binance. 

I just seem to get Market Connectors are not ready.


![Screenshot (81)](https://github.com/hummingbot/hummingbot/assets/69935809/85f06333-0653-4005-a02f-0390aab017bc)

MementoRC commented:
@zethereumGmail @ajjcoppola Ok, something is unexpectedly wrong, It seems that several connections between hummingbot and the connector are missing, for example, even though the user stream reports it is running and subscribing, the strategy does not see it `ready`
It is no use continuing testing, I will update this PR as soon as I get this resolved (you should get notified automatically)
Thanks for the feedback and time

zethereumGmail commented:
I tried running it through VSCode debugger and don't know what I'm doing.

If I can provide anything useful please let me know.

I did the following to get out of Docker and run using the VSCode debugger, it seems to allow me to get a balance but nothing else.


```shell
Windows 10 PowerShell Prompt

wsl --install 

inside wsl
/home/zeth/
git clone https://github.com/hummingbot/hummingbot.git
cd hummingbot
./install
conda activate hummingbot
./compile

git checkout development
git pull
git remote add -f MementoRC https://github.com/MementoRC/hummingbot.git
git config --global user.email "zethereumGmail"
git config --global user.name "Zeth"
git merge MementoRC/master --allow-unrelated-histories
./install
./clean
./compile
bin/hummingbot.py
git pull MementoRC feat/CoinbaseAdvancedTradeV2

bin/hummingbot.py
```

MementoRC commented:
@zethereumGmail , This is all good. There are a few things that are wrong but I can't figure out why. One of the first thing is to get the mapping of trading paris, in the connector I get the correct result, but it does not get passed to hummingbot core and I don't follow why. There is something in the logic of the `ExchangeBase` class that I don't get.

The other thing is that hummingbot does not see the UserStream as ready, but I see that the feed is working (even discarding the first message)

So the connector seems to work, but the handshake with hummingbot is not - Hopefully, I can have a "OMG!, that's so simple!" moment soon - Thanks for your efforts

ajjcoppola commented:
@MementoRC Have my version configured properly now...get a balance on coinbase, looks like it gets the orderbook correctly...just stalls at
await stream_to_pipe_connector
with a repeating "coinbase_advanced_strategy_v2 is not ready" message.

I'm using the modified script: 
buy_only_three_times_example.py
Yet it's not getting to the on_tick() method, where I have a breakpoint.

Let me know how I can help.  Thinking of cutting all the erroneous connects out to otherexchanges, yet not sure that'll help.
OBTW...where on discord do you want to talk about this ...channel name ?

MementoRC commented:
@ajjcoppola Cool! Right, the issue is that for the connector to be ready, it needs to have:
   - trading_pair map initialized (I see the map, but hummingbot core does not - in the ExchangeBase clas which is ... .pyx, not debuggable)
   - the orderbook, it works, but hummingbot core does not know that somehow
   - the user stream, same thing, hummingbot core inspect the WSAssistant _last_recv_time, which i did not set, the implementation uses one WSAssistant per channel/pair, but I will do a all(last_recv for all the ws connections)

So far, the one I don't understand is why the trading pairs are initialized on the Coinbase Advanced Trade side, but not seen on the Hummingbot core side. I changed a few _methods logic flow assuming they were 'private' but they must not have been. I have a bit of a hard-time with Python inheritance, it's too loose for someone who learned OO with the first iterations of C++

We'll get there, it'll require a bit of patience, I am not that experienced programmer.

The `developer-chat` seems to be the best place, I am `Memento 'RC' Mori - Youngling`

MementoRC commented:
@ajjcoppola @zethereumGmail It seems to be due to a flag. The connector registers that the map is initialized internally, but apparently hummingbot resets its reference and place a new call, which fails to call `_update_trading_rules()` since its internal flag is True ... possibly ...

Key change should be:
`coinbase_advanced_trade_v2.py` line 436
```
    async def _initialize_trading_pair_symbol_map(self):
        if not self._pair_symbol_map_initialized:
            await self._update_trading_rules()
            self._pair_symbol_map_initialized: bool = True
```
Also, the UserStream failed to communicate its status, this is also patched

ajjcoppola commented:
@MementoRC Good catch! Looking forward to the new file(s)!  What you have above for the change isn't quite right, as I have that snippet in the code already...Thanks!

MementoRC commented:
@ajjcoppola Sorry for the confusion, yes, the code above needs to be changed by removing the `if` statement. It is in the PR update. Note to always await for the checks to pass (green PR).

ajjcoppola commented:
@MementoRC Got a lot farther..still not verified though...could be my fault though...I'm at the beach till Friday, so I'll monitor this thread, yet won't contribute...you're making good progress though!

![image](https://github.com/hummingbot/hummingbot/assets/4520786/1d45be4e-292d-43f2-b2e6-cce07e2dfb3b)

Here is the script I ran from the scripts dir, with slight changes:
/home/ajjc/hummingbot/scripts/buy_only_three_times_example.py
# --------------------------------------------------------------------------------------
from decimal import Decimal

from hummingbot.client.hummingbot_application import HummingbotApplication
from hummingbot.core.data_type.common import OrderType
from hummingbot.core.event.events import BuyOrderCreatedEvent
from hummingbot.core.rate_oracle.rate_oracle import RateOracle
from hummingbot.strategy.script_strategy_base import ScriptStrategyBase


class BuyOnlyThreeTimesExample(ScriptStrategyBase):
    """
    This example places shows how to add a logic to only place three buy orders in the market,
    use an event to increase the counter and stop the strategy once the task is done.
    """
    order_amount_usd = Decimal(10)
    orders_created = 0
    orders_to_create = 3
    base = "ETH"
    quote = "USD"
    markets = {
        "coinbase_advanced_trade_v2": {f"{base}-{quote}"}
        # "kucoin_paper_trade": {f"{base}-{quote}"}
    }

    def on_tick(self):
        if self.orders_created < self.orders_to_create:
            conversion_rate = RateOracle.get_instance().get_pair_rate(f"{self.base}-USD")
            amount = self.order_amount_usd / conversion_rate
            price = self.connectors["coinbase_advanced_trade_v2"].get_mid_price(f"{self.base}-{self.quote}") * Decimal(0.99)
            self.buy(
                connector_name="coinbase_advanced_trade_v2",
                trading_pair="ETH-USD",
                amount=amount,
                order_type=OrderType.MARKET,
                price=price
            )

    def did_create_buy_order(self, event: BuyOrderCreatedEvent):
        trading_pair = f"{self.base}-{self.quote}"
        if event.trading_pair == trading_pair:
            self.orders_created += 1
            if self.orders_created == self.orders_to_create:
                self.logger().info("All orders created !")
                HummingbotApplication.main_application().stop()

MementoRC commented:
@ajjcoppola @nikspz This one seems tricky. The error message seem to indicate that the exchange order IT is ` ` i.e. empty, which would indicate that it is not returned or translated correctly. From the code, I don't spot anything obvious. We would need to check that the order is listed on the exchange and if it is, find what I missed in interpreting the response

MementoRC commented:
@zethereumGmail I have not implemented support for cancels feed yet, nor did I publish the rate oracle. I will look into these to see if it points to an issue with the connector itself
How long have you let the test srun?
Can you post the logs of the error you observed anyway, they might point to where in the connector a bug lies
Thank you for the help

MementoRC commented:
@nikspz @rapcmia Could you take a look at the coverage, something seems odd. For example, I refactored the `task_manager.py` added several tests but the coverage is still 37%. The `exchange` which has it's v2.1 testing is also very low

Note that I added the RateOracle and the SpotCandles since I think they should be a self-consistent package

The connector is ready for some QA. I ran a few strategies at my end and corrected a few hiccups

nikspz commented:
- Test performed:
    - Cloned and install PR6458 successfully
    - connect coinbase_advaced_trade - got an Error getting all trading pairs from Coinbase Advanced Trade: Error executing request GET https://api.coinbase.com/api/v3/brokerage/products. HTTP status is 502
```
2023-10-20 07:53:57,602 - 3769299 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_exchange.CoinbaseAdvancedTradeExchange - ERROR - Error getting all trading pairs from Coinbase Advanced Trade: Error executing request GET https://api.coinbase.com/api/v3/brokerage/products. HTTP status is 502. Error: N/A
Traceback (most recent call last):
  File "/home/nikita/coinbase6458/hummingbot/connector/exchange/coinbase_advanced_trade/coinbase_advanced_trade_exchange.py", line 620, in _initialize_market_assets
    products: Dict[str, Any] = await self._api_get(
  File "/home/nikita/coinbase6458/hummingbot/connector/exchange_py_base.py", line 863, in _api_get
    return await self._api_request(*args, **kwargs)
  File "/home/nikita/coinbase6458/hummingbot/connector/exchange_py_base.py", line 905, in _api_request
    request_result = await rest_assistant.execute_request(
  File "/home/nikita/coinbase6458/hummingbot/core/web_assistant/rest_assistant.py", line 47, in execute_request
    response = await self.execute_request_and_get_response(
  File "/home/nikita/coinbase6458/hummingbot/core/web_assistant/rest_assistant.py", line 99, in execute_request_and_get_response
    raise IOError(f"Error executing request {method.name} {url}. HTTP status is {response.status}. "
OSError: Error executing request GET https://api.coinbase.com/api/v3/brokerage/products. HTTP status is 502. Error: N/A
```

![image](https://github.com/hummingbot/hummingbot/assets/83953535/a9c5112b-af95-4a22-812a-be86376c67ce)


- Created/started pureMM strategy using coinbase_advaced_trade
- Client generated huge errors log file (1.6 GB)

![image](https://github.com/hummingbot/hummingbot/assets/83953535/e761258c-c800-4538-b2ac-ba8b36d1f216)

created new startegy/restarted Client - unable to start the strategy

![image](https://github.com/hummingbot/hummingbot/assets/83953535/c6078a83-af24-4814-98f6-a7dd107ef889)

checked balance/ restarted strategy:
 - bot was able to place orders
 - Client returns an error Error cancelling orders: Error executing request POST https://api.coinbase.com/api/v3/brokerage/orders/batch_cancel. HTTP status is 502
 - 
 
 
[logs_coin6458.log](https://github.com/hummingbot/hummingbot/files/13052380/logs_coin6458.log)

 
 ```
 2023-10-20 08:48:27,826 - 3898340 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_exchange.CoinbaseAdvancedTradeExchange - ERROR - Error cancelling orders: Error executing request POST https://api.coinbase.com/api/v3/brokerage/orders/batch_cancel. HTTP status is 502. Error: N/A
Traceback (most recent call last):
  File "/home/nikita/coinbase6458/hummingbot/connector/exchange/coinbase_advanced_trade/coinbase_advanced_trade_exchange.py", line 484, in _place_cancels
    cancel_result: Dict[str, Any] = await self._api_post(
  File "/home/nikita/coinbase6458/hummingbot/connector/exchange_py_base.py", line 867, in _api_post
    return await self._api_request(*args, **kwargs)
  File "/home/nikita/coinbase6458/hummingbot/connector/exchange_py_base.py", line 905, in _api_request
    request_result = await rest_assistant.execute_request(
  File "/home/nikita/coinbase6458/hummingbot/core/web_assistant/rest_assistant.py", line 47, in execute_request
    response = await self.execute_request_and_get_response(
  File "/home/nikita/coinbase6458/hummingbot/core/web_assistant/rest_assistant.py", line 99, in execute_request_and_get_response
    raise IOError(f"Error executing request {method.name} {url}. HTTP status is {response.status}. "
OSError: Error executing request POST https://api.coinbase.com/api/v3/brokerage/orders/batch_cancel. HTTP status is 502. Error: N/A
```
![image](https://github.com/hummingbot/hummingbot/assets/83953535/154389a3-e826-4d3b-b637-473d16c2a6de)

there was period bot not connected to coinbase, was able to reconnect

![image](https://github.com/hummingbot/hummingbot/assets/83953535/ec5a6266-56ad-4f7f-9aa2-12cdd90dfc0d)

However not canceled /refreshed orders after 5 minutes (order refresh time is 50s), 
after 5 minutes Client CLI freezed, Duration of the trading session showing same time and not able to interact with client 

![coinbase client crashed](https://github.com/hummingbot/hummingbot/assets/83953535/a59841d2-8bdb-4ecb-930e-76e24fff8718)

MementoRC commented:
@nikspz Thank you for looking into this PR. It is surprising to see so many errors. When looking at `502` HTTP response, it could indicate an issue with the connections, is it possible that the connection to Coinbase was unstable? They had an incident today:
https://status.coinbase.com/incidents/mzp709bj19xh
The log info you posted does not really traceback to the connector, so it's unclear how to investigate. I did debug the `batch_cancel` and it seemed to work. The `products` call as well and you showed you were able to get the balance

nikspz commented:
hi @MementoRC, there's still ongoing issues with orders cancellation.

I think previous orders was not cancelled so DOGE was filled (I had no balance on Doge before)


![image](https://github.com/hummingbot/hummingbot/assets/83953535/0a37bedb-1217-40c0-9a4b-297035dc6af7)

[1b6919a](https://github.com/hummingbot/hummingbot/pull/6458/commits/1b6919a2f0b0c2027e0c920d8add9daf9cf79170)
- Created/started pureMM strategy using `coinbase_advanced_trade` 
- Review logs flooded with `Error cancelling orders`: Error executing request POST https://api.coinbase.com/api/v3/brokerage/orders/batch_cancel. HTTP status is 502. Error: N/A
- Review WARNING - `Failed to update balances`. Error: Error executing request GET https://api.coinbase.com/api/v3/brokerage/accounts. HTTP status is 503. Error: go/sg/75aeb9b9-f54d-4067-a082-5a6e447428b5

[1b6919a-buyandsell.zip](https://github.com/hummingbot/hummingbot/files/13072261/1b6919a-buyandsell.zip)
[logs_pure_cb.log](https://github.com/hummingbot/hummingbot/files/13072264/logs_pure_cb.log)

After orders refreshed (it takes around 24 minutes), no balance available for SELL and BUY orders both
![image](https://github.com/hummingbot/hummingbot/assets/83953535/ca168b25-2216-4eb5-9c1c-849a604f9921)

Restarted with Debug:
![image](https://github.com/hummingbot/hummingbot/assets/83953535/8618bbc7-c050-4290-8a94-ba205dee6fa9)


Here on sell events, Client have balance to place sell order, but failed to place it:


[1b6919a.zip](https://github.com/hummingbot/hummingbot/files/13072252/1b6919a.zip)
[logs_coinbase_sell.log](https://github.com/hummingbot/hummingbot/files/13072212/logs_coinbase_sell.log)
```
2023-10-23 14:43:04,001 - 86903 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (DOGE-USD) Creating 1 bid orders at (Size, Price): ['0.2 DOGE, 0.06309 USD']
2023-10-23 14:43:04,002 - 86903 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (DOGE-USD) Creating 1 ask orders at (Size, Price): ['6E+1 DOGE, 0.06373 USD']
2023-10-23 14:43:04,002 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_exchange.CoinbaseAdvancedTradeExchange - WARNING - Buy order amount 0.2 is lower than the minimum order size 3.8. The order will not be created, increase the amount to be higher than the minimum order size.
2023-10-23 14:43:04,013 - 86903 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1698072184.0, "order_id": "CBAT-BDEUD608633d5496686965ab97f", "order_type": "OrderType.LIMIT_MAKER", "event_name": "MarketOrderFailureEvent", "event_source": "coinbase_advanced_trade"}
2023-10-23 14:43:04,014 - 86903 - hummingbot.connector.client_order_tracker - INFO - Order CBAT-BDEUD608633d5496686965ab97f has failed. Order Update: OrderUpdate(trading_pair='DOGE-USD', update_timestamp=1698072184.0, new_state=<OrderState.FAILED: 6>, client_order_id='CBAT-BDEUD608633d5496686965ab97f', exchange_order_id=None, misc_updates=None)
2023-10-23 14:43:04,304 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_exchange.CoinbaseAdvancedTradeExchange - NETWORK - Error submitting sell LIMIT_MAKER order to Coinbase_advanced_trade for 60.0 DOGE-USD 0.06373.
Traceback (most recent call last):
  File "/home/nikita/coinbase6458/hummingbot/connector/exchange_py_base.py", line 452, in _create_order
    await self._place_order_and_process_update(order=order, **kwargs,)
  File "/home/nikita/coinbase6458/hummingbot/connector/exchange/coinbase_advanced_trade/coinbase_advanced_trade_exchange.py", line 343, in _place_order_and_process_update
    exchange_order_id, update_timestamp = await self._place_order(
  File "/home/nikita/coinbase6458/hummingbot/connector/exchange/coinbase_advanced_trade/coinbase_advanced_trade_exchange.py", line 339, in _place_order
    raise ValueError(f"Failed to place order on {self.name}. Error: {order_result['error_response']}")
ValueError: Failed to place order on coinbase_advanced_trade. Error: {'error': 'INVALID_LIMIT_PRICE_POST_ONLY', 'message': '', 'error_details': '', 'preview_failure_reason': 'PREVIEW_INVALID_LIMIT_PRICE_POST_ONLY'}
2023-10-23 14:43:04,305 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_exchange.CoinbaseAdvancedTradeExchange - WARNING - Failed to submit SELL order to Coinbase_advanced_trade. Check API key and network connection.
2023-10-23 14:43:04,310 - 86903 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1698072184.0, "order_id": "CBAT-SDEUD608633d54982b6965ab97f", "order_type": "OrderType.LIMIT_MAKER", "event_name": "MarketOrderFailureEvent", "event_source": "coinbase_advanced_trade"}
2023-10-23 14:43:04,311 - 86903 - hummingbot.connector.client_order_tracker - INFO - Order CBAT-SDEUD608633d54982b6965ab97f has failed. Order Update: OrderUpdate(trading_pair='DOGE-USD', update_timestamp=1698072184.0, new_state=<OrderState.FAILED: 6>, client_order_id='CBAT-SDEUD608633d54982b6965ab97f', exchange_order_id=None, misc_updates=None)
2023-10-23 14:43:13,447 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_api_user_stream_data_source.CoinbaseAdvancedTradeAPIUserStreamDataSource - DEBUG - Sequence handler heartbeats:395
2023-10-23 14:43:13,447 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_api_user_stream_data_source.CoinbaseAdvancedTradeAPIUserStreamDataSource - DEBUG - {'channel': 'heartbeats', 'client_id': '', 'timestamp': '2023-10-23T14:43:13.363118791Z', 'sequence_num': 395, 'events': [{'current_time': '2023-10-23 14:43:13.362521689 +0000 UTC m=+425760.164580852', 'heartbeat_counter': 14191}]}
2023-10-23 14:43:13,448 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_api_user_stream_data_source.CoinbaseAdvancedTradeAPIUserStreamDataSource - DEBUG - Sequence handler heartbeats:150
2023-10-23 14:43:13,448 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_api_user_stream_data_source.CoinbaseAdvancedTradeAPIUserStreamDataSource - DEBUG - {'channel': 'heartbeats', 'client_id': '', 'timestamp': '2023-10-23T14:43:13.362842145Z', 'sequence_num': 150, 'events': [{'current_time': '2023-10-23 14:43:13.362521689 +0000 UTC m=+425760.164580852', 'heartbeat_counter': 14191}]}
2023-10-23 14:43:13,466 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_api_user_stream_data_source.CoinbaseAdvancedTradeAPIUserStreamDataSource - DEBUG - Sequence handler heartbeats:158
2023-10-23 14:43:13,466 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_api_user_stream_data_source.CoinbaseAdvancedTradeAPIUserStreamDataSource - DEBUG - {'channel': 'heartbeats', 'client_id': '', 'timestamp': '2023-10-23T14:43:13.362762774Z', 'sequence_num': 158, 'events': [{'current_time': '2023-10-23 14:43:13.362521689 +0000 UTC m=+425760.164580852', 'heartbeat_counter': 14191}]}
2023-10-23 14:43:22,777 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_api_user_stream_data_source.CoinbaseAdvancedTradeAPIUserStreamDataSource - DEBUG - Sequence handler heartbeats:157
2023-10-23 14:43:22,777 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_api_user_stream_data_source.CoinbaseAdvancedTradeAPIUserStreamDataSource - DEBUG - {'channel': 'heartbeats', 'client_id': '', 'timestamp': '2023-10-23T14:43:22.695015845Z', 'sequence_num': 157, 'events': [{'current_time': '2023-10-23 14:43:22.694630618 +0000 UTC m=+425760.255532568', 'heartbeat_counter': 14191}]}
2023-10-23 14:43:43,387 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_api_user_stream_data_source.CoinbaseAdvancedTradeAPIUserStreamDataSource - DEBUG - Sequence handler heartbeats:396
2023-10-23 14:43:43,387 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_api_user_stream_data_source.CoinbaseAdvancedTradeAPIUserStreamDataSource - DEBUG - {'channel': 'heartbeats', 'client_id': '', 'timestamp': '2023-10-23T14:43:43.307532871Z', 'sequence_num': 396, 'events': [{'current_time': '2023-10-23 14:43:43.30748841 +0000 UTC m=+425790.109547574', 'heartbeat_counter': 14192}]}
2023-10-23 14:43:43,388 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_api_user_stream_data_source.CoinbaseAdvancedTradeAPIUserStreamDataSource - DEBUG - Sequence handler heartbeats:151
2023-10-23 14:43:43,388 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_api_user_stream_data_source.CoinbaseAdvancedTradeAPIUserStreamDataSource - DEBUG - {'channel': 'heartbeats', 'client_id': '', 'timestamp': '2023-10-23T14:43:43.307735695Z', 'sequence_num': 151, 'events': [{'current_time': '2023-10-23 14:43:43.30748841 +0000 UTC m=+425790.109547574', 'heartbeat_counter': 14192}]}
2023-10-23 14:43:43,393 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_api_user_stream_data_source.CoinbaseAdvancedTradeAPIUserStreamDataSource - DEBUG - Sequence handler heartbeats:159
2023-10-23 14:43:43,393 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_api_user_stream_data_source.CoinbaseAdvancedTradeAPIUserStreamDataSource - DEBUG - {'channel': 'heartbeats', 'client_id': '', 'timestamp': '2023-10-23T14:43:43.307693094Z', 'sequence_num': 159, 'events': [{'current_time': '2023-10-23 14:43:43.30748841 +0000 UTC m=+425790.109547574', 'heartbeat_counter': 14192}]}
2023-10-23 14:43:47,166 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_exchange.CoinbaseAdvancedTradeExchange - ERROR - Error cancelling orders: Error executing request POST https://api.coinbase.com/api/v3/brokerage/orders/batch_cancel. HTTP status is 502. Error: N/A
Traceback (most recent call last):
  File "/home/nikita/coinbase6458/hummingbot/connector/exchange/coinbase_advanced_trade/coinbase_advanced_trade_exchange.py", line 482, in _place_cancels
    cancel_result: Dict[str, Any] = await self._api_post(
  File "/home/nikita/coinbase6458/hummingbot/connector/exchange_py_base.py", line 867, in _api_post
    return await self._api_request(*args, **kwargs)
  File "/home/nikita/coinbase6458/hummingbot/connector/exchange_py_base.py", line 905, in _api_request
    request_result = await rest_assistant.execute_request(
  File "/home/nikita/coinbase6458/hummingbot/core/web_assistant/rest_assistant.py", line 47, in execute_request
    response = await self.execute_request_and_get_response(
  File "/home/nikita/coinbase6458/hummingbot/core/web_assistant/rest_assistant.py", line 99, in execute_request_and_get_response
    raise IOError(f"Error executing request {method.name} {url}. HTTP status is {response.status}. "
OSError: Error executing request POST https://api.coinbase.com/api/v3/brokerage/orders/batch_cancel. HTTP status is 502. Error: N/A
2023-10-23 14:43:52,774 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_api_user_stream_data_source.CoinbaseAdvancedTradeAPIUserStreamDataSource - DEBUG - Sequence handler heartbeats:158
2023-10-23 14:43:52,775 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_api_user_stream_data_source.CoinbaseAdvancedTradeAPIUserStreamDataSource - DEBUG - {'channel': 'heartbeats', 'client_id': '', 'timestamp': '2023-10-23T14:43:52.694979409Z', 'sequence_num': 158, 'events': [{'current_time': '2023-10-23 14:43:52.694777055 +0000 UTC m=+425790.255679005', 'heartbeat_counter': 14192}]}
2023-10-23 14:44:13,385 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_api_user_stream_data_source.CoinbaseAdvancedTradeAPIUserStreamDataSource - DEBUG - Sequence handler heartbeats:397
2023-10-23 14:44:13,386 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_api_user_stream_data_source.CoinbaseAdvancedTradeAPIUserStreamDataSource - DEBUG - {'channel': 'heartbeats', 'client_id': '', 'timestamp': '2023-10-23T14:44:13.308164304Z', 'sequence_num': 397, 'events': [{'current_time': '2023-10-23 14:44:13.307602122 +0000 UTC m=+425820.109661286', 'heartbeat_counter': 14193}]}
2023-10-23 14:44:13,391 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_api_user_stream_data_source.CoinbaseAdvancedTradeAPIUserStreamDataSource - DEBUG - Sequence handler heartbeats:152
2023-10-23 14:44:13,391 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_api_user_stream_data_source.CoinbaseAdvancedTradeAPIUserStreamDataSource - DEBUG - {'channel': 'heartbeats', 'client_id': '', 'timestamp': '2023-10-23T14:44:13.307899028Z', 'sequence_num': 152, 'events': [{'current_time': '2023-10-23 14:44:13.307602122 +0000 UTC m=+425820.109661286', 'heartbeat_counter': 14193}]}
2023-10-23 14:44:13,396 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_api_user_stream_data_source.CoinbaseAdvancedTradeAPIUserStreamDataSource - DEBUG - Sequence handler heartbeats:160
2023-10-23 14:44:13,397 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_api_user_stream_data_source.CoinbaseAdvancedTradeAPIUserStreamDataSource - DEBUG - {'channel': 'heartbeats', 'client_id': '', 'timestamp': '2023-10-23T14:44:13.307874468Z', 'sequence_num': 160, 'events': [{'current_time': '2023-10-23 14:44:13.307602122 +0000 UTC m=+425820.109661286', 'heartbeat_counter': 14193}]}
2023-10-23 14:44:15,002 - 86903 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (DOGE-USD) Creating 1 bid orders at (Size, Price): ['6E+1 DOGE, 0.06309 USD']
2023-10-23 14:44:15,002 - 86903 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (DOGE-USD) Creating 1 ask orders at (Size, Price): ['6E+1 DOGE, 0.06373 USD']
2023-10-23 14:44:15,323 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_exchange.CoinbaseAdvancedTradeExchange - NETWORK - Error submitting sell LIMIT_MAKER order to Coinbase_advanced_trade for 60.0 DOGE-USD 0.06373.
Traceback (most recent call last):
  File "/home/nikita/coinbase6458/hummingbot/connector/exchange_py_base.py", line 452, in _create_order
    await self._place_order_and_process_update(order=order, **kwargs,)
  File "/home/nikita/coinbase6458/hummingbot/connector/exchange/coinbase_advanced_trade/coinbase_advanced_trade_exchange.py", line 343, in _place_order_and_process_update
    exchange_order_id, update_timestamp = await self._place_order(
  File "/home/nikita/coinbase6458/hummingbot/connector/exchange/coinbase_advanced_trade/coinbase_advanced_trade_exchange.py", line 339, in _place_order
    raise ValueError(f"Failed to place order on {self.name}. Error: {order_result['error_response']}")
ValueError: Failed to place order on coinbase_advanced_trade. Error: {'error': 'INVALID_LIMIT_PRICE_POST_ONLY', 'message': '', 'error_details': '', 'preview_failure_reason': 'PREVIEW_INVALID_LIMIT_PRICE_POST_ONLY'}
2023-10-23 14:44:15,324 - 86903 - hummingbot.connector.exchange.coinbase_advanced_trade.coinbase_advanced_trade_exchange.CoinbaseAdvancedTradeExchange - WARNING - Failed to submit SELL order to Coinbase_advanced_trade. Check API key and network connection.
2023-10-23 14:44:15,328 - 86903 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1698072255.0, "order_id": "CBAT-SDEUD60863418ff9a26965ab97f", "order_type": "OrderType.LIMIT_MAKER", "event_name": "MarketOrderFailureEvent", "event_source": "coinbase_advanced_trade"}
```

![image](https://github.com/hummingbot/hummingbot/assets/83953535/770a4160-3661-4506-b292-d46ba105adaa)

MementoRC commented:
@nikspz Confirming 502 errors are definitively a Coinbase side issue, their forum has lots of people reporting the same issue: https://forums.coinbasecloud.dev/t/502-bad-gateway/5406/11

ajjcoppola commented:
MementoRC,
Good catch!
alan
On 10/23/2023 1:50:54 PM, MementoRC ***@***.***> wrote:
@nikspz [https://github.com/nikspz] Confirming 502 errors are definitively a Coinbase side issue, their forum has lots of people reporting the same issue: https://forums.coinbasecloud.dev/t/502-bad-gateway/5406/11 [https://forums.coinbasecloud.dev/t/502-bad-gateway/5406/11]
—
Reply to this email directly, view it on GitHub [https://github.com/hummingbot/hummingbot/pull/6458#issuecomment-1775999266], or unsubscribe [https://github.com/notifications/unsubscribe-auth/ABCPWUWEMKCTZVALTR6ZMNDYA3KC5AVCNFSM6AAAAAA2DZVHPKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTONZVHE4TSMRWGY].
You are receiving this because you were mentioned.Message ID: ***@***.***>

MementoRC commented:
@nikspz Thank you.
   - What should we do if a LIMIT_MAKER is rejected by the exchange? Initially, I thought it was a rare event, but this time you got another one right away, so raising a `ValueError` does not seem appropriate. Could we resubmit as LIMIT, or continue by returning an UNKNOWN for the order_id

cardosofede commented:
@MementoRC this is ready for dev review?

MementoRC commented:
@cardosofede It's close. There seems to be a memory leak somewhere that I have not found

nikspz commented:
hi @MementoRC Could you please add updates on this bounty PR?

MementoRC commented:
@nikspz, @ajjcoppola has been using it and doing some testing to try to figure out the remaining issues that he seem to be experiencing with the WS. On their forum, Coinbase have had several posts on their WS and at my end I have not found issues in the code.

I just browsed their Changelog and noticed that they had now an REST time call that I should use instead of their `Sign-In` API from v2, may be the time-sync will be improved

cardosofede commented:
@MementoRC @nikspz @rapcmia I think that this PR is here for a long time and Memento did a lot of work.
Memento do you think that you can clean up the last changes that you want to do next week and we can merge it so other people can start testing it?
I saw that there is a problem regarding to WS, the disconnection and reconnection of Hummingbot is not enough?

nikspz commented:
@MementoRC Could you please fix the conflicts for us to proceed?

MementoRC commented:
@nikspz I resolved the conflicts. @cardosofede I will try to update the time server next week. I have been tied-up with other projects, but will take a look next week
Thx

MementoRC commented:
@ajjcoppola @nikspz @cardosofede I think I found the issue with the orders, I now register `Insuffucient Funds` as OrderState.FAILED

ajjcoppola commented:
Let me try my tests, hopefully work now!

nikspz commented:
Tests:
- Cloned and installed branch commit 731c05eb276eaf6aadf115e7f1c3f56f2dcaecbd
- connect coinbase_advanced_trade failed
- failed to check balance
- Failed to get pairs from coinbase
- Failed strategy creation checks

![image](https://github.com/hummingbot/hummingbot/assets/83953535/a3f1f4ca-3820-4563-8fe4-3abe9cacfd88)
![image](https://github.com/hummingbot/hummingbot/assets/83953535/f0132fdc-6d21-4bb1-8ac1-8a18a1465dd6)
![balance coinbase 731c05eb276eaf6aadf115e7f1c3f56f2dcaecbd](https://github.com/hummingbot/hummingbot/assets/83953535/6b5bc08d-0c94-460d-991e-13164fd4dd78)
![image](https://github.com/hummingbot/hummingbot/assets/83953535/4bcd0e77-a8af-412c-b399-02779c4f8198)
![image](https://github.com/hummingbot/hummingbot/assets/83953535/a728e16d-268a-432b-ba08-6adf167eec94)

MementoRC commented:
@nikspz The connector status shows `Keys NOT confirmed`. The new time server requires authentication. I have just pushed an update on the time server, I had not noticed that the API payload had changed

nikspz commented:
Steps:
- Cloned and installed branch commit 01a7bed
- connect coinbase_advanced_trade failed

[logs_hummingbot.log](https://github.com/hummingbot/hummingbot/files/14431697/logs_hummingbot.log)


![connect coinbase 228 failed](https://github.com/hummingbot/hummingbot/assets/83953535/eebf791f-be3d-4ff7-93f9-aba62ae54ff6)

MementoRC commented:
@nikspz Do you mind retrying to connect with the additional debug info. This is odd, @ajjcoppola is able to connect and trade, I think

ajjcoppola commented:
@nikspz Yes, I've been using my coinbase account to test the connector for months now.  The only thing I can think might be different is that I'm using the dev branch from @MementoRC, which I keep fully up to date with so that I can access the connector code...mystery...

fengtality commented:
@MementoRC We discussed this pull request at length during our meeting this morning, and have decided to close this pull request and unassign you from this bounty.

In addition to the testing issues encountered above, this approach taken by this PR is very different from the current [Spot Connector Standard](https://hummingbot.org/developers/connectors/). I understand that there may be Coinbase-specific functionality that needs to be handled, but this connector introduces may deviations from the standard that will not be maintainable by anyone by you in the future. Since Hummingbot is a community-maintained codebase, we need to ensure that all connectors conform to a standard that allows any experienced Hummingbot developer to maintain them.

Afterwards, we will re-scope this bounty using the [Bounty Pricing Guidelines](https://hummingbot.org/bounties/prices/).
